docs/job-queue.md

```diff
@@ -0,0 +1,154 @@
+
+The compile benchmark suite for a given target is split into `N` benchmark sets. To run the suite, `N` collectors (with benchmark sets `0, 1, ..., N`) have to be available.
+
+Each collector has a hard-coded benchmark set that it always benchmarks. Benchmarks should ideally not move within the sets, to ensure that each benchmark will always be benchmarked on the same machine, to avoid unnecessary environment noise.
+
+That being said, sometimes it might be useful to balance the sets a little bit, to ensure that all collectors can run their jobs in approximately the same duration, to avoid unbalanced workloads.
+
+The fact that the benchmark sets are assigned to collectors *statically* and there is no load balancing or work-stealing going on in-between the sets means that if one of the collectors stops running, **it will halt the whole system**. In that case, a [manual intervention](./deployment.md) might be required.
+
+#### Backfilling
+When the website enqueues jobs of a request, it also enqueues the jobs with the same parameters for its parent request. Parents should always be benchmarked before their children, so in most cases, all the parent jobs will already be present in the job queue.
+
+However, someone can create a try request with *non-default* parameters. For example, they could request benchmarking `cranelift` codegen backend, which is not normally benchmarked on master requests. When that happens, we need to ensure that we will also run those non-default jobs for the *parent* in addition to the try request itself, otherwise we wouldn't have anything to compare to.
+
+This situation is called `backfilling`. When the website enqueues jobs for a request with non-default parameters, it will create *new jobs* also for its parent request, which did not exist before. The collectors will then go and also benchmark those parent jobs, thus "backfilling" the results into a request that was already completed previously. The status of the parent request does not change when this happens, it stays `Completed` or `Failed`.
+
+## Collectors
+
+The main job of each collector is to continuously dequeue jobs from the job queue, run all of their benchmarks and store the results into the database.
+
+### Registration
+Individual collectors have to be registered in the database, so that we can show their status on the [status page], even if they were offline at the moment.
+
+Their information is stored in the `collector_config` table. Each collector has its assigned benchmark set and target, which are used to determine which jobs the collector will handle.
+
+To register a new collector, you can run the following command:
+
+```bash
+cargo run --bin collector add_collector \
+  --collector_name "<name>" \
+  --target <target> \
+  --benchmark_set "<benchmark-set>" \
+  --is_active
```

Comment:
Presumably this needs to be run with credentials - maybe on the instance itself?